Remotion/Web/Core/Context/HttpContextStorageProviderNetCore.cs

```diff
@@ -0,0 +1,182 @@
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Web;
+using Remotion.Context;
+using Remotion.ServiceLocation;
+using Remotion.Utilities;
+
+namespace Remotion.Web.Context
+{
+#if !NETFRAMEWORK
+  [ImplementationFor(typeof(ISafeContextStorageProvider), Position = 0)]
+  public class HttpContextStorageProvider : ISafeContextStorageProvider, ISafeContextBoundaryStrategy
+  {
+    private struct HashTableAccessor
+    {
+      private class WrappedHashtable
+      {
+        public HttpContext Context { get; }
+        public Hashtable Data { get; }
+
+        public WrappedHashtable (HttpContext context, Hashtable data)
+        {
+          Context = context;
+          Data = data;
+        }
+      }
+
+      private const string c_dataDictionaryKey = "Rm.DD";
+
+      private readonly ThreadLocal<WeakReference<WrappedHashtable?>> _dataTable;
+
+      public HashTableAccessor (bool _)
+      {
+        _dataTable = new(() => new WeakReference<WrappedHashtable?>(null));
+      }
+
+      public bool TryGetCachedHashtable (HttpContext httpContext, [MaybeNullWhen(false)] out Hashtable hashtable)
+      {
+        if (_dataTable.Value!.TryGetTarget(out var target) && target.Context == httpContext)
+        {
+          hashtable = target.Data;
+          return true;
+        }
+
+        var wrappedHashtable = (WrappedHashtable?)httpContext.Items[c_dataDictionaryKey];
+        hashtable = wrappedHashtable?.Data;
+
+        if (wrappedHashtable != null)
+        {
+          _dataTable.Value!.SetTarget(wrappedHashtable);
+          return hashtable != null;
+        }
+
+        return false;
+      }
+
+      public Hashtable GetOrCreateHashtable (HttpContext context)
+      {
+        if (TryGetCachedHashtable(context, out var hashtable))
+        {
+          return hashtable;
+        }
+
+        return SetHashtable(context, new Hashtable());
+      }
+
+      public Hashtable? RemoveHashtable (HttpContext context)
+      {
+        if (TryGetCachedHashtable(context, out var cachedHashtable))
+        {
+          context.Items.Remove(c_dataDictionaryKey);
+          _dataTable.Value!.SetTarget(null!);
+        }
+
+        return cachedHashtable;
+      }
+
+      public Hashtable SetHashtable (HttpContext context, Hashtable hashtable)
+      {
+        var wrappedHashtable = new WrappedHashtable(context, hashtable);
+        context.Items[c_dataDictionaryKey] = wrappedHashtable;
+        _dataTable.Value!.SetTarget(wrappedHashtable);
+        return hashtable;
+      }
+    }
+
+    // Ignore the obsolete warnings for AsyncLocalStorageProvider
+#pragma warning disable 618
+    private readonly AsyncLocalStorageProvider _fallbackProvider = new();
+#pragma warning restore 618
+
+    private readonly HashTableAccessor _hashTableAccessor = new(true);
+
+    public SafeContextBoundary OpenSafeContextBoundary ()
+    {
+      var httpContext = HttpContext.Current;
+
+      if (httpContext == null)
```

Comment:
style: I personally dislike the empty line in this case and would write it without. Otherwise the `httpContext` feels so left alone :(